flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingStrategy.java

```diff
@@ -40,9 +40,9 @@
 	/**
 	 * Called whenever vertices need to be restarted (due to task failure).
 	 *
-	 * @param verticesToRestart The tasks need to be restarted
+	 * @param verticesToRestart The tasks to be restarted, topologically sorted
 	 */
-	void restartTasks(Set<ExecutionVertexID> verticesToRestart);
+	void restartTasks(List<ExecutionVertexID> verticesToRestart);
```

Comment:
What is the benefit if the SchedulingStrategy sees the vertices in topological order?

Reply:
This makes it easier for vertices to `#allocateSlotsAndDeploy` vertices in topological order.

But you remind me that we can do it in a different way that do not change this interface: having a set of vertices to schedule, iterate over all the vertices in the topology in topological order to create an ordered list, and then invoke `#allocateSlotsAndDeploy` on the list.

Reply:
This PR is now made in another way, which significantly reduced the changes especially on production codes.
The basic idea is that in `DefaultScheduler` we do not sort the vertices anymore but respect the order given by the scheduling strategy.
The scheduling strategy is now responsible for ordering the vertices to schedule if it feels like to. This should be an easy work thought since it can get to know the topological order from the `SchedulingTopology`.